crates/astria-sequencer/src/grpc/state_ext.rs

```diff
@@ -272,124 +107,230 @@ pub(crate) trait StateReadExt: StateRead {
         // 5. block hash to rollup transactions proof
         // 6. block hash to rollup IDs proof
 
-        let key = block_hash_by_height_key(block.height().into());
-        self.nonverifiable_put_raw(key, block.block_hash().to_vec());
-
-        let rollup_ids = block
-            .rollup_transactions()
-            .keys()
-            .copied()
-            .map(From::from)
-            .collect::<Vec<_>>();
-
-        let key = rollup_ids_by_hash_key(&block.block_hash());
-
-        self.nonverifiable_put_raw(
-            key,
-            borsh::to_vec(&RollupIdSeq(rollup_ids))
-                .wrap_err("failed to serialize rollup IDs list")?,
-        );
-
-        let key = sequencer_block_header_by_hash_key(&block.block_hash());
         let SequencerBlockParts {
             block_hash,
             header,
             rollup_transactions,
             rollup_transactions_proof,
             rollup_ids_proof,
         } = block.into_parts();
-        let header = header.into_raw();
-        self.nonverifiable_put_raw(key, header.encode_to_vec());
 
-        for (id, rollup_data) in rollup_transactions {
-            let key = rollup_data_by_hash_and_rollup_id_key(&block_hash, &id);
-            self.nonverifiable_put_raw(key, rollup_data.into_raw().encode_to_vec());
-        }
+        put_block_hash(self, header.height(), block_hash)?;
+        put_rollup_ids(self, &block_hash, rollup_transactions.keys().copied())?;
+        put_block_header(self, &block_hash, header)?;
+        put_rollups_transactions(self, &block_hash, rollup_transactions.into_iter())?;
+        put_rollups_transactions_proof(self, &block_hash, rollup_transactions_proof)?;
+        put_rollup_ids_proof(self, &block_hash, rollup_ids_proof)
+    }
+}
 
-        let key = rollup_transactions_proof_by_hash_key(&block_hash);
-        self.nonverifiable_put_raw(key, rollup_transactions_proof.into_raw().encode_to_vec());
+#[instrument(skip_all)]
+async fn get_sequencer_block_by_hash<S: StateRead + ?Sized>(
+    state: &S,
+    hash: &[u8; 32],
+) -> Result<SequencerBlock> {
+    let header = state
+        .get_sequencer_block_header_by_hash(hash)
+        .await
+        .wrap_err("failed to get sequencer block header by hash")?;
+    let rollup_ids = state
+        .get_rollup_ids_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup ids by block hash")?;
+    let rollup_transactions_proof = state
+        .get_rollup_transactions_proof_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup transactions proof by block hash")?;
+    let rollup_ids_proof = state
+        .get_rollup_ids_proof_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup ids proof by block hash")?;
+
+    #[expect(
+        clippy::default_trait_access,
+        reason = "want to avoid explicitly importing `index_map` crate to sequencer crate"
```

Comment:
Builder to the rescue. :)

Reply:
A good constructor would be even better! One which took an iterator over `RollupTransactions` so we don't need to care how it's held internally :)